wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DevfileEnvironmentProvisioner.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.factory.server.urlfactory;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.util.Collections.emptyMap;
+import static org.eclipse.che.api.devfile.server.Constants.KUBERNETES_TOOL_TYPE;
+import static org.eclipse.che.api.devfile.server.Constants.OPENSHIFT_TOOL_TYPE;
+
+import java.util.Optional;
+import java.util.function.Function;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+import org.eclipse.che.commons.annotation.Nullable;
+
+/** Creates workspace environment from specific tool in devfile if any. */
+@Singleton
+public class DevfileEnvironmentProvisioner {
+
+  static final String DEFAULT_RECIPE_CONTENT_TYPE = "application/x-yaml";
+
+  private final URLFetcher urlFetcher;
+
+  @Inject
+  public DevfileEnvironmentProvisioner(URLFetcher urlFetcher) {
+    this.urlFetcher = urlFetcher;
+  }
+
+  /**
+   * Finds an recipe-type tool (openshift or kubernetes) in devfile and tries to provision {@link
+   * EnvironmentImpl} from it. If such tools are present in devfile, an file URL composer function
```

Comment:
> ... If such tools are present in devfile ...

I guess it must be only `ONE` or `0`. Maybe it would be better to throw an exception in case when there are multiply `Kubernetes`/`OpenShift` tools instead of ignoring them?

Reply:
It is covered by validator, but yes maybe it makes sense.